daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala

```diff
@@ -120,4 +120,48 @@ object Util {
       )
   }
 
+  private[this] def toInterface(choice: AbstractTemplateChoice[_]): AbstractTemplateChoice[Unit] =
+    choice match {
+      case AbstractTemplateChoice(name, consuming, _, selfBinder, argBinder, returnType, _) =>
+        AbstractTemplateChoice(name, consuming, (), selfBinder, argBinder, returnType, ())
+    }
+
+  private[this] def toInterface(key: AbstractTemplateKey[_]): AbstractTemplateKey[Unit] =
+    key match {
+      case AbstractTemplateKey(typ, _, _) =>
+        AbstractTemplateKey(typ, (), ())
+    }
+
+  private[this] def toInterface(template: AbstractTemplate[_]): AbstractTemplate[Unit] =
+    template match {
+      case AbstractTemplate(param, _, _, _, choices, _, key) =>
+        AbstractTemplate(
+          param,
+          (),
+          (),
+          (),
+          choices.transform((_, v) => toInterface(v)),
+          (),
+          key.map(toInterface),
+        )
+    }
+
+  def toInterface(p: Package): PackageInterface = {
+    val moduleInterface = p.modules.transform((_, mod) =>
+      mod.collectDefinitions {
+        case DDataType(true, params, cons) =>
```

Comment:
So this only gives us the LF interface that we also need for codegens. This seems enough for typechecking of ledger API values but as you pointed out it’s not enough for validation. So I assume for ledgers this means that we still end up retaining the whole package in memory? (hopefully we can at least get rid of it in triggers now)

Reply:
It is replace by `toSignature` that include all type information.